design/keycloak.x/storage-persistence.md

```diff
@@ -0,0 +1,530 @@
+# Keycloak X - Storage / Persistence layer
+
+* **Status**: Draft
+* **JIRA**: https://issues.redhat.com/browse/KEYCLOAK-7301
+
+## Goals of this document
+
+This documents aims to address the following items:
+
+* Define a strategy for no-downtime storage upgrades
+* State the requirements imposed on the persistent part of the store
+* Provide design of layer that allows for simple yet performant
+  implementations of custom stores
+
+## Non-Goals of this document
+
+This documents aims to _not_ address the following items:
+
+* Provide exact prescription of how the data are stored in the data store
+* Discuss design of caching or its impact on performance
+* Discuss pros and cons of individual storage technologies
+
+## Introduction
+
+In an ideal world, Keycloak cluster upgrade would be done seamlessly -
+no service disruption, new version rolled out without user noticing.
+
+This can only be done if all relevant components support this.
+This document specifically discusses persistence aspect. Other aspects,
+e.g. caching or API compatibility, are left for other specifications.
+
+There are several strategies of updates with respect to persistence possible.
+Those that do not work are listed below:
+
+*  _Stop the world._ Stop old version; Update all the data; Start new version.
+
+   This is the current state of Keycloak. It disrupts the service
+   and thus is not an option.
+
+* _[Blue-green deployment.](https://martinfowler.com/bliki/BlueGreenDeployment.html)_
+  Two production-grade clusters with a router that redirects traffic
+  to the updated one once update is done.
+
+  This is problematic for Keycloak since it maintains state in the cluster.
+  For example, code-to-token flow could be disrupted upon switching the traffic.
+
+  The same concern applies to a variant of blue-green deployment
+  where only a portion of traffic is directed to the updated
+  cluster and the rest is directed to the old one, with the proportions
+  being gradually changed towards the new cluster.
+
+* _Keep the store schema unchanged._ Update the nodes gradually
+  to the new version.
+
+  With the current database (as of Keycloak 11), this would work
+  for most of the maintenance releases. However it would
+  not work if there were new features that introduce more
+  than cosmetic changes the database.
+
+The last option is attractive though. Could the store structural changes
```

Comment:
Probably a nitpick, but it may not be 100% clear what exactly means "structural" changes and what is "change in object models" ? The latter is described below though, but at the point when newbie person is reading this, it may be a bit confused. But maybe it is just me :-)

Reply:
I rephrased it in the new version. Thank you for the review!